ponix4-src/usr/share/man/man8/acpi_listen.8
.TH acpi_listen 8 "Nov 2003"
.\" Portions Copyright (c) 2003 Sun Microsystems
.\" Copyright (c) 2004 Tim Hockin (thockin@hockin.org)
.\" Some parts (C) 2003 - Gismo / Luca Capello <luca.pca.it> http://luca.pca.it
.SH NAME
acpi_listen \- ACPI event listener
.SH SYNOPSIS
\fBacpi_listen\fP [\fIoptions\fP]

.SH DESCRIPTION
\fBacpid\fP is the system-wide ACPI event catcher.  \fBacpi_listen\fP is a
simple shell-friendly tool which connects to acpid and listens for events.
When an event occurs, acpi_listen will print it on stdout.

.SH OPTIONS
.TP
.BI \-c "\fR, \fP" \--count " events"
Receive up to a maximum number of ACPI events, then exit.
.TP
.BI \-s "\fR, \fP" \--socketfile " filename"
This option changes the name of the UNIX domain socket which \fBacpid\fP opens.
Default is \fI/var/run/acpid.socket\fP.
.TP
.BI \-t "\fR, \fP" \--time " seconds"
Listen for the specified time in seconds, before exiting.
.TP
.BI \-v "\fR, \fP" \--version
Print version information and exit.
.TP
.BI \-h "\fR, \fP" \--help
Show help and exit.

.SH FILES
.PD 0
.B /var/run/acpid.socket
.PD

.SH BUGS
There are no known bugs.  To file bug reports, see \fBAUTHORS\fP below.
.SH SEE ALSO
acpid(8)
.SH AUTHORS
Ted Felix (www.tedfelix.com)
.br
Tim Hockin <thockin@hockin.org>
.br
Luca Capello <luca@pca.it>
.br



ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/acpid-2.0.23/acpi_listen.c
/*
 *  acpi_listen.c - ACPI client for acpid's UNIX socket
 *
 *  Portions Copyright (C) 2003 Sun Microsystems (thockin@sun.com)
 *  Some parts (C) 2003 - Gismo / Luca Capello <luca.pca.it> http://luca.pca.it
 *  Copyright (C) 2004 Tim Hockin (thockin@hockin.org)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <getopt.h>
#include <time.h>
#include <sys/poll.h>
#include <grp.h>
#include <signal.h>

#include "acpid.h"
#include "ud_socket.h"

static int handle_cmdline(int *argc, char ***argv);
static char *read_line(int fd);

const char *progname;
const char *socketfile = ACPID_SOCKETFILE;
static int max_events;

static void
time_expired(int signum __attribute__((unused)))
{
	exit(EXIT_SUCCESS);
}

int
main(int argc, char **argv)
{
	int sock_fd;
	int ret;

	/* handle an alarm */
	signal(SIGALRM, time_expired);

	/* learn who we really are */
	progname = (const char *)strrchr(argv[0], '/');
	progname = progname ? (progname + 1) : argv[0];

	/* handle the commandline  */
	handle_cmdline(&argc, &argv);

	/* open the socket */
	sock_fd = ud_connect(socketfile);
	if (sock_fd < 0) {
		fprintf(stderr, "%s: can't open socket %s: %s\n",
			progname, socketfile, strerror(errno));
		exit(EXIT_FAILURE);
	}
	/* set stdout to be line buffered */
	setvbuf(stdout, NULL, _IOLBF, 0);

	/* main loop */
	ret = 0;
	while (1) {
		char *event;

		/* read and handle an event */
		event = read_line(sock_fd);
		if (event) {
			fprintf(stdout, "%s\n", event);
		} else if (errno == EPIPE) {
			fprintf(stderr, "connection closed\n");
			break;
		} else {
			static int nerrs;
			if (++nerrs >= ACPID_MAX_ERRS) {
				fprintf(stderr, "too many errors - aborting\n");
				ret = 1;
				break;
			}
		}

		if (max_events > 0 && --max_events == 0) {
			break;
		}
	}

	return ret;
}

static struct option opts[] = {
	{"count", 0, 0, 'c'},
	{"socketfile", 1, 0, 's'},
	{"time", 0, 0, 't'},
	{"version", 0, 0, 'v'},
	{"help", 0, 0, 'h'},
	{NULL, 0, 0, 0},
};
static const char *opts_help[] = {
	"Set the maximum number of events.",	/* count */
	"Use the specified socket file.",	/* socketfile */
	"Listen for the specified time (in seconds).",/* time */
	"Print version information.",		/* version */
	"Print this message.",			/* help */
};

static void
usage(FILE *fp)
{
	struct option *opt;
	const char **hlp;
	int max, size;

	fprintf(fp, "Usage: %s [OPTIONS]\n", progname);
	max = 0;
	for (opt = opts; opt->name; opt++) {
		size = strlen(opt->name);
		if (size > max)
			max = size;
	}
	for (opt = opts, hlp = opts_help; opt->name; opt++, hlp++) {
		fprintf(fp, "  -%c, --%s", opt->val, opt->name);
		size = strlen(opt->name);
		for (; size < max; size++)
			fprintf(fp, " ");
		fprintf(fp, "  %s\n", *hlp);
	}
}

/*
 * Parse command line arguments
 */
static int
handle_cmdline(int *argc, char ***argv)
{
	for (;;) {
		int i;
		i = getopt_long(*argc, *argv, "c:s:t:vh", opts, NULL);
		if (i == -1) {
			break;
		}
		switch (i) {
		case 'c':
			if (!isdigit(optarg[0])) {
				usage(stderr);
				exit(EXIT_FAILURE);
			}
			max_events = atoi(optarg);
			break;
		case 's':
			socketfile = optarg;
			break;
		case 't':
			if (!isdigit(optarg[0])) {
				usage(stderr);
				exit(EXIT_FAILURE);
			}
			alarm(atoi(optarg));
			break;
		case 'v':
			printf(PACKAGE "-" VERSION "\n");
			exit(EXIT_SUCCESS);
		case 'h':
			usage(stdout);
			exit(EXIT_SUCCESS);
		default:
			usage(stderr);
			exit(EXIT_FAILURE);
			break;
		}
	}

	*argc -= optind;
	*argv += optind;

	return 0;
}

#define MAX_BUFLEN	1024
static char *
read_line(int fd)
{
	static char *buf;
	int buflen = 64;
	int i = 0;
	int r;
	int searching = 1;

	while (searching) {
		buf = realloc(buf, buflen);
		if (!buf) {
			fprintf(stderr, "ERR: malloc(%d): %s\n",
				buflen, strerror(errno));
			return NULL;
		}
		memset(buf+i, 0, buflen-i);

		while (i < buflen) {
			r = TEMP_FAILURE_RETRY (read(fd, buf+i, 1) );
			if (r < 0) {
				/* we should do something with the data */
				fprintf(stderr, "ERR: read(): %s\n",
					strerror(errno));
				return NULL;
			} else if (r == 0) {
				/* signal this in an almost standard way */
				errno = EPIPE;
				return NULL;
			} else if (r == 1) {
				/* scan for a newline */
				if (buf[i] == '\n') {
					searching = 0;
					buf[i] = '\0';
					break;
				}
				i++;
			}
		}
		if (buflen >= MAX_BUFLEN) {
			break;
		}
		buflen *= 2;
	}

	return buf;
}
